Refactor to leverage PaymentGatewayCommon instead of BroadleafCommon

ckittrell opened the pull request:
**A Brief Overview**
This refactoring is being done to divorce the payment gateway implementation from BLC monolith-specific patterns, concepts, assumptions, etc. & make the module more generally applicable, regardless of where/how it's being used.

**Link to MicroPM issue**
https://github.com/BroadleafCommerce/MicroPM/issues/1298

**Additional context**
NOTE: This is defining the initial state of the `develop-ms-1.0.x` branch which will be maintained entirely separately from the monolith-based `develop-x.x.x` branches & releases.

PR depends on: https://github.com/BroadleafCommerce/PaymentGatewayCommon/pull/3

ckittrell committed:
Refactor to leverage the PaymentGatewayCommon library & resolve simple compilation issues

ckittrell committed:
- Update pom to reflect microservices patterns & leverage PaymentGatewayCommon
- Fix compilation issues with the usage of PaymentGatewayCommon instead of BraodleafCommon

https://github.com/BroadleafCommerce/MicroPM/issues/1298

ckittrell committed:
- Refactor bean registration to use annotation-based patterns & AutoConfiguration classes
- Refactor resource injection to use constructor injection
- Rename implementations to use "Default*" instead of "*Impl"
- Remove the BroadleafPayPalCheckoutController to avoid monolith-based assumptions

https://github.com/BroadleafCommerce/MicroPM/issues/1298

ckittrell committed:
Refactor PayPal service names

https://github.com/BroadleafCommerce/MicroPM/issues/1298

ckittrell committed:
Introduce NonNullApi package-level annotations

https://github.com/BroadleafCommerce/MicroPM/issues/1298

ckittrell committed:
Simple spotless cleanup

https://github.com/BroadleafCommerce/MicroPM/issues/1298

ckittrell committed:
- Update version to 1.0.0-SNAPSHOT since this artifact is now under the "com.broadleafcommerce.microservices" groupId
- Introduce Maven & Jenkins build updates

https://github.com/BroadleafCommerce/MicroPM/issues/1298

ckittrell committed:
- Update docs to use asciidoc instead of markdown

https://github.com/BroadleafCommerce/MicroPM/issues/1298

ckittrell committed:
Refactor PayPal service names to include "PayPalCheckout" which is a reference to the PayPal implementation

https://github.com/BroadleafCommerce/MicroPM/issues/1298

ckittrell committed:
- Update Jenkinsfile to include the docs directory

https://github.com/BroadleafCommerce/MicroPM/issues/1298

phillipuniverse commented on pom.xml:
```diff
@@ -135,5 +91,9 @@
 
+        <dependency>
+            <groupId>org.projectlombok</groupId>
+            <artifactId>lombok</artifactId>
```
This is _required_ to be an `optional` dependency it should _never_ come through transitively. We really should be validating that, I added it as a task here: https://github.com/BroadleafCommerce/jenkins-pipeline-library/issues/21

phillipuniverse commented on pom.xml:
```diff
@@ -109,24 +69,20 @@
             <artifactId>rest-api-sdk</artifactId>
             <version>1.14.0</version>
         </dependency>
+
```
Remove the servlet API dependency version. It is already defined in the parent. It is possible that you will get an error about not having a version here. I can't remember when exactly the groupId/artifactId coordinates changed, but you might have to change the groupId/artifactId coordinates to work with servlet 3.0+

phillipuniverse commented on docs/pom.xml:
```diff
@@ -0,0 +1,84 @@
+                        <groupId>org.apache.maven.plugins</groupId>
+                        <artifactId>maven-antrun-plugin</artifactId>
+                    </plugin>
+                    <!-- Copy over the generated snippets from the main module -->
```
This entire section is a copy/pasta problem.

phillipuniverse commented on docs/pom.xml:
```diff
@@ -0,0 +1,84 @@
+                        <artifactId>asciidoctor-maven-plugin</artifactId>
+                        <configuration>
+                            <attributes>
+                                <snippets>${project.build.outputDirectory}/generated-snippets</snippets>
```
Also a copy/paste bug

phillipuniverse commented on src/main/java/org/broadleafcommerce/payment/service/gateway/DefaultPayPalCheckoutConfiguration.java:
```diff
@@ -1,107 +1,91 @@
 
-    @Autowired
-    protected SystemPropertiesService propertiesService;
+    private final Environment env;
```
I'm struggling with what we should do here. On the one hand, it would be nice to maintain the ability to merge changes up into this branch from the legacy branch, in which case we should leave the code similar to how it was before. On the other hand, the way we do properties has changed a lot with our microservices patterns, and specifically how those properties are documented. For example, what I want to say is that you should refactor _all_ of the properties in this entire module to use `@ConfigurationProperties` instead and add Javadocs to each of those properties indicating what they do and how they are used instead of relying on the `Environment`.

@elbertbautista what do you think we should do here?

ckittrell replied:
Either route has a bad side unfortunately. I lean towards refactoring to use `@ConfigurationProperties` so that the available configuration is clearly communicated to customers, but I'm not thrilled about the potential for tricky merging situations. 

At the same time, this isn't the only item that would lead to difficult merging. That leads me to:
- What other changes would cause a merge to be challenging? - updated package structures, refactoring of the `PaymentGatewayServiceProvider`, etc.
- How much more of a challenge would this change add?

phillipuniverse replied:
>  I lean towards refactoring to use @ConfigurationProperties so that the available configuration is clearly communicated to customers, but I'm not thrilled about the potential for tricky merging situations.

Perhaps we should just treat this as it's _not_ a merge at all. Instead, the previous version will continue to exist exactly as it has in perpetuity, but in maintenance mode. This version of the library will evolve and change where _it_ makes sense.

Coming back to this I think this is the right answer and continued mergability is not really that much of a concern. @elbertbautista?

ckittrell replied:
That sounds right to me.

... leaving this conversation unresolved to give @elbertbautista an opportunity to object.

phillipuniverse commented on src/main/java/org/broadleafcommerce/payment/service/gateway/autoconfigure/PayPalServiceAutoConfiguration.java:
```diff
@@ -0,0 +1,106 @@
+
+    @Bean
+    @ConditionalOnMissingBean
+    public PaymentGatewayConfigurationService paymentGatewayConfigurationService(
```
I think that the original idea of the payment gateways was that you could have multiple payment gateways in a single project. Especially for Paypal; it is common to accept Paypal accounts and use a different provider for credit cards (maybe Stripe, idk).

However, the way you have this auto configuration you have effectively done this:

```java
@ConditionalOnMissingBean(PaymentGatewayConfigurationService.class)
```

as `@ConditionalOnMissingBean` uses the method return value to check if you do not provide a manual value  in the annotation itself. So, this means that if some other service produces a `PaymentGatewayConfigurationService` (which every single one will) then this bean will not be registered.

@elbertbautista can you clarify the requirements here on what needs to hold true for these payment gateways?

phillipuniverse replied:
It's also likely that we use some different patterns than we had before in the payment gateway service in regards to how these beans are used.

ckittrell replied:
@phillipuniverse great catch! It's definitely still expected that multiple gateways will be used within a single project. The pattern used to register these beans needs to be updated.

phillipuniverse reviewed (changes requested):
The biggest unknown I have at this point is what to do with these configuration properties and all places we use `Environment` and _especially_ `@Value`
- Not sure if we should change packages or not

phillipuniverse commented:
I added the start of what will be required for Jenkins at https://github.com/BroadleafCommerce/devops/issues/14

ckittrell committed:
- Update Lombok dependency to be optional
- Update javax servlet-api dependency to leverage the version provided by the parent

https://github.com/BroadleafCommerce/MicroPM/issues/1298

ckittrell committed:
- Remove old release note docs

https://github.com/BroadleafCommerce/MicroPM/issues/1298

ckittrell committed:
- Refactor project structure so that it includes a parent pom with "services" & "docs" sub-modules

https://github.com/BroadleafCommerce/MicroPM/issues/1298

ckittrell committed:
- Implement getGatewayType() methods for each gateway service
- Introduce "PayPalCheckout*" service interfaces to provide a clean mechanism for the auto configuration to register the "ConditionalOnMissingBean" beans without interfering with any other payment gateway

https://github.com/BroadleafCommerce/MicroPM/issues/1298

ckittrell committed:
- Update externalCallService param name

https://github.com/BroadleafCommerce/MicroPM/issues/1298

ckittrell committed:
- Remove usages of abstract classes that have been removed from PaymentGatewayCommon
- Fix compilation issues due to dependencies that were made optional in PaymentGatewayCommon

https://github.com/BroadleafCommerce/MicroPM/issues/1298

phillipuniverse commented:
Ok so there is still an open question on what to do with properties but everything else in here is probably fine.

I added docs on [moving builds to new Jenkins](https://github.com/BroadleafCommerce/jenkins-pipeline-library/blob/master/README.adoc#moving-builds-from-old-jenkins-to-new-jenkins) which will need to happen on all of the branches that have a `Jenkinsfile`:

- develop-2.7.x
- develop-2.8.x
- develop-3.0.x

This new branch that you'll merging into will be based on a slightly different pipeline library but the old pipeline library now builds on new Jenkins too. We will need to apply similar treatment on all of our other libraries (everything will eventually move to new Jenkins) but we can move them as we go.

ckittrell committed:
- Refactor the resolution of properties
- Remove a few presentation properties that should no longer be responsibility of this backend service
- Split API credentials/configuration from the PaymentGatewayConfiguration implementation
- Update usages of returnUrl/cancelUrl to leverage the PaymentRequest to gather the siteBaseUrl instead of expecting it to be resolved via a filter & be generally accessible (old monolith pattern)

https://github.com/BroadleafCommerce/MicroPM/issues/1298

ckittrell commented:
@phillipuniverse I went ahead & made the property resolution/documentation changes. To go along with those changes, I also created the following PR for PaymentGatewayCommon: (https://github.com/BroadleafCommerce/PaymentGatewayCommon/pull/4).

phillipuniverse commented on services/src/main/java/org/broadleafcommerce/payment/service/gateway/PayPalCheckoutRestConfiguration.java:
```diff
@@ -0,0 +1,173 @@
+ * @author Elbert Bautista (elbertbautista)
+ * @author Chris Kittrell
+ */
+@ConfigurationProperties("gateway.paypal.checkout.rest")
```
Since you're changing this already, change the property names to start with `broadleaf`, e.g. `braodleaf.paypal.checkout.rest`

phillipuniverse commented on services/src/main/java/org/broadleafcommerce/payment/service/gateway/PayPalCheckoutRestConfiguration.java:
```diff
@@ -0,0 +1,173 @@
+    private String totalType = MessageConstants.TOTAL;
+
+    /**
+     * <p>
```
Do not use any Javadoc formatting it does not come through. You can test this by doing autocompletion of properties in your IDE, any Javadoc formatting does not show up correctly.

phillipuniverse commented on services/src/main/java/org/broadleafcommerce/payment/service/gateway/PayPalCheckoutRestConfiguration.java:
```diff
@@ -0,0 +1,173 @@
+ */
+@ConfigurationProperties("gateway.paypal.checkout.rest")
+@Data
+public class PayPalCheckoutRestConfiguration {
```
Yeah remove all the HTML and Javadoc tags from all the properties here.

ckittrell committed:
- Update PayPal rest API properties to begin with 'broadleaf' instead of 'gateway'

https://github.com/BroadleafCommerce/MicroPM/issues/1298

ckittrell committed:
- Update configuration properties class name
- Add spring-boot-configuration-processor dependency so that a spring-configuration-metadata.json file is produced

https://github.com/BroadleafCommerce/MicroPM/issues/1298
